core/base/array.cpp

```diff
@@ -92,9 +94,42 @@ void Array<ValueType>::fill(const ValueType value)
 }
 
 
+template <typename ValueType>
+void reduce(const Array<ValueType>& input_arr, ValueType* value)
+{
+    auto exec = input_arr.get_executor();
+    exec->run(array::make_reduce_array(input_arr.get_const_data(),
+                                       input_arr.get_num_elems(), value));
```

Comment:
value needs to be the device pointer, right?
if only the pointer in the public interface, we do not know whether it is on the device or not.

Reply:
Yes, the idea was that if you needed the value on the host, then you would use the reduce function which returns a value. Yes, I think this would throw a seg-fault, but I did not want to wrap this in an Array again.